### PR TITLE
[SPARK-42555][CONNECT][FOLLOWUP] Add the new proto msg to support the remaining jdbc API

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
https://github.com/apache/spark/pull/40252 supported some jdbc API that reuse the proto msg `DataSource`. The `DataFrameReader` also have another kind jdbc API that is unrelated to load data source.

### Why are the changes needed?
This PR adds the new proto msg `PartitionedJDBC` to support the remaining jdbc API.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 1]
```diff
@@ -250,6 +250,47 @@ class DataFrameReader private[sql] (sparkSession: SparkSession) extends Logging
...
+   *
+   * @param table
+   *   Name of the table in the external database.
+   * @param predicates
+   *   Condition in the where clause for each partition.
+   * @param connectionProperties
+   *   JDBC database connection arguments, a list of arbitrary string tag/value. Normally at least
+   *   a "user" and "password" property should be included. "fetchsize" can be used to control the
+   *   number of rows per fetch.
+   * @since 1.4.0
```
Comment (zhengruifeng) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:274]:
```suggestion
   * @since 3.4.0
```

**Comment by beliefer** [comment]
ping @hvanhovell @HyukjinKwon @dongjoon-hyun  cc @LuciferYang

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 2]
```diff
@@ -140,6 +141,21 @@ message Read {
...
+
+  message PartitionedJDBC {
+    // (Required) JDBC URL.
+    string url = 1;
+
+    // (Required) Name of the table in the external database.
+    string table = 2;
+
+    // (Optional) Condition in the where clause for each partition.
+    repeated string predicates = 3;
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:153]: Can we just put the predicates into the DataSource message?

Comment (beliefer) [reply connector/connect/common/src/main/protobuf/spark/connect/relations.proto:154]: But the transform path is very different from DataSource.

Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/relations.proto:154]: I think it's simple to add a `if-else` in `transformReadRel`, if we can reuse existing DataSource message (with new field  `predicates `)

Comment (beliefer) [reply connector/connect/common/src/main/protobuf/spark/connect/relations.proto:154]: OK. Let's put the predicates into the DataSource message.

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 3]
```diff
@@ -140,6 +140,9 @@ message Read {
 
     // (Optional) A list of path for file-system backed data sources.
     repeated string paths = 4;
+
+    // (Optional) Condition in the where clause for each partition.
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:144]: Please add the comment that this currently only works for jdbc.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 4]
```diff
@@ -250,6 +250,46 @@ class DataFrameReader private[sql] (sparkSession: SparkSession) extends Logging
...
+   *   a "user" and "password" property should be included. "fetchsize" can be used to control the
+   *   number of rows per fetch.
+   * @since 3.4.0
+   */
+  def jdbc(
+      url: String,
+      table: String,
+      predicates: Array[String],
+      connectionProperties: Properties): DataFrame = {
+    sparkSession.newDataFrame { builder =>
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:281]: Can you please set the format to JDBC? We are now relying the presence of predicates to figure out that something is a JDBC table. That is relying far too heavily on the client doing the right thing, for example what would happen if you set format = parquet and still define predicates?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 5]
```diff
@@ -684,26 +686,43 @@ class SparkConnectPlanner(val session: SparkSession) {
...
-        if (rel.getDataSource.hasSchema && rel.getDataSource.getSchema.nonEmpty) {
-
-          DataType.parseTypeWithFallback(
-            rel.getDataSource.getSchema,
-            StructType.fromDDL,
-            fallbackParser = DataType.fromJson) match {
-            case s: StructType => reader.schema(s)
-            case other => throw InvalidPlanInput(s"Invalid schema $other")
+
+        if (rel.getDataSource.getPredicatesCount == 0) {
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:691]: Please make the logic a bit like this:
```scala
if (format == "jdbc" && rel.getDataSource.getPredicatesCount) {
  // Plan JDBC with predicates
} else id (rel.getDataSource.getPredicatesCount == 0) {
 // Plan datasource
} else {
  throw InvalidPlan(s"Predicates are not supported for $format datasources.)"
}

```

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 4 above)
Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:281]: Yeah. we can't rely on client.

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto (hunk 3 above)
Comment (beliefer) [reply connector/connect/common/src/main/protobuf/spark/connect/relations.proto:144]: OK

**Comment by beliefer** [comment]
@hvanhovell Do you have any other advice? cc @HyukjinKwon @zhengruifeng @dongjoon-hyun

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by beliefer** [comment]
@hvanhovell @zhengruifeng Thank you.

**Pull request closed without merging**